src/master-playlist-controller.js

```diff
@@ -831,40 +847,81 @@ export default class MasterPlaylistController extends videojs.EventTarget {
       let segment = media.segments[i];
 
-      if ('cueOut' in segment || 'cueOutCont' in segment || 'cueIn' in segment) {
-        let cueJson = {};
+      if (cue !== undefined) {
 
-        if ('cueOut' in segment) {
-          cueJson.cueOut = segment.cueOut;
+        if ('cueIn' in segment) {
+          // Found a CUE-IN so end the cue
+          if (mediaTime !== cue.endTime) {
+            // If the end time does not match the current mediaTime, a CUE-IN may have
+            // been inserted earlier than expected, so the cue should be updated.
+            cue.endTime = mediaTime;
+            cue.adEndTime = mediaTime;
+          }
+          mediaTime += segment.duration;
+          cue = undefined;
+          continue;
         }
-        if ('cueOutCont' in segment) {
-          cueJson.cueOutCont = segment.cueOutCont;
+
+        if (mediaTime < cue.endTime) {
+          // Already processed this mediaTime for this cue
+          mediaTime += segment.duration;
+          continue;
         }
-        if ('cueIn' in segment) {
-          cueJson.cueIn = segment.cueIn;
+
+        // otherwise extend cue until a CUE-IN is found
+        cue.endTime += segment.duration;
+
+      } else {
+        // Since the cues will span for at least the segment duration, adding a fudge
+        // factor of half segment duration will prevent duplicate cues from being
+        // created when timing info is not exact (e.g. cue start time initialized
+        // at 10.006677, but next call mediaTime is 10.003332 )
+        cue = findAdCue(track, mediaTime + (segment.duration / 2));
+        if (cue) {
+          // there is a cue already created for this mediaTime
+          // decrement i and not increase mediaTime to handle this segment
+          // again with this cue
+          i--;
+          continue;
+        }
+
+        if ('cueOut' in segment) {
+          cue = new window.VTTCue(mediaTime,
+                                  mediaTime + segment.duration,
+                                  '');
+          cue.adStartTime = mediaTime;
+          cue.adEndTime = mediaTime + parseFloat(segment.cueOut);
+          track.addCue(cue);
         }
 
-        // Use a short duration for the cue point, as it should trigger for a segment
-        // transition (in this case, defined as the beginning of the segment that the tag
-        // precedes), but keep it for a minimum of 0.5 seconds to remain usable (won't
-        // lose it as an active cue by the time a user retrieves the active cues).
-        this.cueTagsTrack_.addCue(new window.VTTCue(mediaTime,
-                                                    mediaTime + 0.5,
-                                                    JSON.stringify(cueJson)));
+        if ('cueOutCont' in segment) {
+          // Entered into the middle of an ad cue
+          let adOffset;
+          let adTotal;
+
+          [adOffset, adTotal] = segment.cueOutCont.split('/').map((value) => {
```

Comment:
I think its ok to make the assumption for now since the parsing only affects `adStartTime` and `adEndTime` which don't actually affect the functionality of the cue, but only give some extra metadata to the user. `parseFloat` will also ignore anything after it sees the first non-numerical character, so as long as that data is first, it won't be affected.